README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# bingmapr

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![License:
MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
<!-- badges: end -->

bingmapr is an R packge for retrieving and plotting maps from the [Bing
Static Maps
API](https://docs.microsoft.com/en-us/bingmaps/rest-services/imagery/get-a-static-map).
I created this package in May 2021 primarily to to access Bing’s Bird’s
Eye (oblique-angle) imagery so the routes and traffic features of the
API are not currently supported. There are no immediate plans to add
these features to the package but pull requests are welcome.

This package is inspired by the [snapbox
package](https://github.com/anthonynorth/snapbox/) that allows the use
of the maps from the Mapbox static map API in ggplot maps. I hope to
integrate bingmapr with ggplot in the future.

You can also explore the static map API in the browser using the [Bing
Static Map Maker](https://staticmapmaker.com/bing/) made by [Katy
DeCorah](https://katydecorah.com/).

## Installation

You can install the development version of bingmapr from
[GitHub](https://github.com/) with:

``` r
# install.packages("pak")
pak::pkg_install("elipousson/bingmapr")
```

## Example

This package requires a Bing Maps API Key which you can get here:
<https://www.bingmapsportal.com/>

``` r
bing_maps_api_key(key = "put your key here", install = TRUE)
```

Here are examples you can run with an installed key:

``` r
# Get request URL centered on Washington Square Park in New York, NY
get_request_url(location = c(40.7308349, -73.99746074), orientation = "W")

# Get an sf object to show how bingmapr can use sf objects for locations
nc <- sf::st_read(system.file("shape/nc.shp", package = "sf"))

# Plot a static map showing the center of Wake County, North Carolina
get_map_image(location = nc[nc$NAME == "Wake", ], orientation = 180)

# Plot static map centered 300 meters north and 500 meters west of the center
get_map_image(location = nc[nc$NAME == "Wake", ], orientation = 180, nudge = c(300, 500))
```


NEWS.md
<!-- NEWS.md is maintained by https://cynkra.github.io/fledge, do not edit -->

# bingmapr 0.1.0.9000 (2022-10-18)

- add support for bbox and sfc locations to `get_request_url()`
- remove `plot_map_image()` by incorporating `magick::image_read()` into `get_map_image()`
- add `sf_to_coords()` and `nudge_location()` helper functions
- drop jpeg and graphics package dependencies
- `get_map_meta()` function now optionally returns a bbox from the metadata




R/utils.R
#' @noRd
#' @importFrom sf st_centroid st_coordinates st_transform st_as_sfc st_union
sf_to_coords <- function(location) {
  if (inherits(location, "bbox")) {
    location <- sf::st_as_sfc(location)
  }

  location <- suppressWarnings(sf::st_centroid(sf::st_union(location)))
  location <- sf::st_coordinates(sf::st_transform(location, 4326))
  rev(c(location))
}

#' @noRd
nudge_location <- function(location, nudge = NULL) {
  nudge_lat <- nudge[1]
  nudge_lon <- nudge[2]

  lat <- location[1]
  lon <- location[2]

  # from https://stackoverflow.com/questions/7477003/calculating-new-longitude-latitude-from-old-n-meters
  # 6371000.0 is approximate radius of the earth in meters
  new_lat <- lat + (nudge_lat / 6371000.0) * (180 / pi)
  new_lon <- lon + (nudge_lon / 6371000.0) * (180 / pi) / cos(lat * pi / 180)

  c(new_lat, new_lon)
}


#' Get closest cardinal direction based on angle orientation
#' @noRd
closest_cardinal_dir <- function(orientation = NULL) {
  cardinal_dirs <- c(0, 45, 135, 225, 360)

  cardinal_dirs[
    findInterval(
      orientation,
      cardinal_dirs
    )
  ]
}


R/bing_maps_api_key.R
#' Install a Bing Maps API Key in Your '.Renviron' File for Repeated Use
#'
#' This function will add your CENSUS API key to your .Renviron file so it can
#' be called securely without being stored in your code. After you have
#' installed your key, it can be called any time by typing
#' `Sys.getenv("BING_MAPS_API_KEY")`.
#'
#' @param key The API key provided to you from the Bing Maps Dev Center
#'   formatted in quotes. A key can be acquired at
#'   <https://www.bingmapsportal.com/>
#' @param overwrite If this is set to `TRUE`, it will overwrite an existing
#'   BING_MAPS_API_KEY that you already have in your .Renviron file.
#' @param install if `TRUE`, will install the key in your .Renviron file for use
#'   in future sessions. Defaults to `FALSE`
#' @rdname bing_maps_api_key
#' @export
#' @importFrom utils read.table write.table
bing_maps_api_key <- function(key, overwrite = FALSE, install = FALSE) {
  if (install) {
    home <- Sys.getenv("HOME")
    renv <- file.path(home, ".Renviron")
    if (file.exists(renv)) {
      file.copy(renv, file.path(home, ".Renviron_backup"))
    }
    if (!file.exists(renv)) {
      file.create(renv)
    } else {
      if (isTRUE(overwrite)) {
        message("Your original .Renviron will be backed up and stored in your R HOME directory if needed.")
        oldenv <- utils::read.table(renv, stringsAsFactors = FALSE)
        newenv <- oldenv[-grep(
          "BING_MAPS_API_KEY",
          oldenv
        ), ]
        utils::write.table(newenv, renv,
          quote = FALSE, sep = "\n",
          col.names = FALSE, row.names = FALSE
        )
      } else {
        tv <- readLines(renv)
        if (any(grepl("BING_MAPS_API_KEY", tv))) {
          stop("A BING_MAPS_API_KEY already exists. You can overwrite it with the argument overwrite=TRUE",
            call. = FALSE
          )
        }
      }
    }
    keyconcat <- paste0("BING_MAPS_API_KEY='", key, "'")
    write(keyconcat, renv, sep = "\n", append = TRUE)
    message("Your API key has been stored in your .Renviron and can be accessed by Sys.getenv(\"BING_MAPS_API_KEY\"). \nTo use now, restart R or run `readRenviron(\"~/.Renviron\")`")
    return(key)
  } else {
    message("To install your API key for use in future sessions, run this function with `install = TRUE`.")
    Sys.setenv("BING_MAPS_API_KEY" = key)
  }
}

#' @noRd
get_bing_maps_api_key <- function(key = NULL) {
  if (is.null(key)) {
    key <- Sys.getenv("BING_MAPS_API_KEY")
  }

  stopifnot(
    "A valid Bing Maps API key is required" = is.character(key) && (key != "")
  )

  key
}


R/map_image.R
.onAttach <- function(libname, pkgname) {
  packageStartupMessage("Use of the Bing Maps APIs is governed by the Microsoft Bing Maps Platform APIs Terms Of Use.\nPlease visit https://www.microsoft.com/en-us/maps/product/ for more information.")
}

#' @noRd
imagery_options <-
  c(
    "Aerial", "AerialWithLabels", "AerialWithLabelsOnDemand",
    "Streetside",
    "BirdsEye", "BirdsEyeWithLabels",
    "Road",
    "CanvasDark", "CanvasLight", "CanvasGray"
  )

#' Get and plot Bing Static Maps
#'
#' See the documentation on Bing Static Maps for reference:
#' <https://docs.microsoft.com/en-us/bingmaps/rest-services/imagery/get-a-static-map>
#'
#' Get API key from <https://www.bingmapsportal.com/>
#'
#' @param location A `sf`, `sfc`, or `bbox` object (centroid point is used for
#'   location) or numeric vector in format `c(latitude, longitude)`. Defaults to
#'   `NULL`.
#' @param query String with query for location. query is ignored if a location
#'   is provided.  Defaults to `NULL`.
#' @param imagery String with imagery type, Default: 'BirdsEye' Supported values
#'   include:
#' - Aerial: Aerial imagery.
#' - AerialWithLabels: Aerial imagery with a road overlay.
#' - AerialWithLabelsOnDemand: Aerial imagery with on-demand road overlay.
#' - Streetside: Street-level imagery.
#' - BirdsEye: Birds Eye (oblique-angle) imagery.
#' - BirdsEyeWithLabels: Birds Eye (oblique-angle) imagery with a road overlay.
#' - Road: Roads without additional imagery.
#' - CanvasDark: A dark version of the road maps.
#' - CanvasLight: A lighter version of the road maps which also has some of the
#' details such as hill shading disabled.
#' - CanvasGray: A grayscale version of the road maps
#' @param width,height,mapsize Width and height in pixels or use mapsize to
#'   provide a vector of c(width, height). If mapsize is provided, width and
#'   height are ignored. Default: 600px width, 400px height, mapsize is `NULL`.
#' @param zoom Numeric vector between 0 and 20 for imagery other than Bird's Eye
#'   maps or 18 to 22 for Bird's Eye maps. Default: 18
#' @param orientation Orientation as a character string ("N", "E", "S", "W") or
#'   length 1 numeric vector (0,90,180,270). Other numeric orientations (from
#'   -360 to 720) are matched to the closest value, e.g. 35 to 0 or 75 to 90.
#'   Default: 0
#' @param nudge Numeric vector in the format, `c(meters to shift latitude,
#'   meters to shift longitude)`, e.g. `c(100, 0)` to shift center 100 meters in
#'   latitude.  Defaults to `NULL`.
#' @param bbox If `TRUE`, return a `bbox` class object based on the bounding box
#'   values from the map metadata. If `FALSE` (default), return the full JSON
#'   metadata. ([get_map_meta()] only)
#' @param key Bing Maps API Key, Default: `Sys.getenv("BING_MAPS_API_KEY")`
#' @param check If `TRUE`, check the map metadata which returns an error if the
#'   image is unavailable. Default `FALSE`.
#' @param .perform If `TRUE`, return results from [httr2::req_perform()]. If
#'   `FALSE`, return request.
#' @param ... Additional parameters passed to [httr2::req_url_query()]
#' @inheritParams magick::image_read
#' @name bing_static_map
#' @md
NULL

#' @return get_request_url returns the request URL for the Static Map API
#' @rdname bing_static_map
#' @export
#' @importFrom httr2 request req_url_path_append req_url_query req_user_agent
#'   req_error resp_body_json req_perform
req_bingmapr <- function(location = NULL,
                         query = NULL,
                         imagery = "BirdsEye",
                         width = 600,
                         height = 400,
                         mapsize = NULL,
                         zoom = 18,
                         orientation = 0,
                         nudge = NULL,
                         key = Sys.getenv("BING_MAPS_API_KEY"),
                         check = FALSE,
                         .perform = TRUE,
                         ...) {
  req <-
    httr2::request(
      "https://dev.virtualearth.net/REST/V1/Imagery/Map"
    )

  imagery <-
    match.arg(
      imagery,
      imagery_options
    )

  req <-
    httr2::req_url_path_append(
      req,
      imagery,
      get_location_or_query(location, query, nudge, imagery),
      get_zoom_for_imagery(zoom, imagery)
    )

  req <-
    httr2::req_url_query(
      req,
      dir = get_dir_from_orientation(orientation),
      ms = get_ms_from_mapsize(width, height, mapsize),
      key = get_bing_maps_api_key(key),
      ...
    )

  req <-
    httr2::req_user_agent(
      req,
      "bingmapr (https://github.com/elipousson/bingmapr)"
    )

  if (check) {
    req <-
      httr2::req_error(
        req,
        body = function(resp) {
          details <- httr2::resp_body_json(resp)$errorDetails[[1]]
          if (details == "The zoom level is not between 0 and 22, or there is no Birdseye imagery at the specified location.") {
            return("There is no Birdseye imagery at the specified location.")
          } else {
            details
          }
        }
      )
  }


  if (!.perform) {
    return(req)
  }

  httr2::req_perform(req)
}


#' Convert numeric or character vector for orientation into dir query parameter
#' @noRd
get_location_or_query <- function(location = NULL,
                                  query = NULL,
                                  nudge = NULL,
                                  imagery = NULL) {
  if (is.null(location)) {
    if (is.character(query) && !(imagery %in% imagery_options[5:6])) {
      location <- query
    } else {
      stop(
        paste0(
          "location must be provided to use the Bird's Eye imagery types.",
          "\nThe query parameter is not supported."
        )
      )
    }
  } else {
    if (inherits(location, c("sfc", "sf", "bbox"))) {
      location <- sf_to_coords(location)
    }
    stopifnot(
      "location must be a numeric coordinate pair or a sf, sfc, or bbox object." = is.numeric(location)
    )

    if (is.numeric(nudge)) {
      location <- nudge_location(location, nudge)
    }
  }

  paste0(location, collapse = ",")
}

#' @noRd
get_ms_from_mapsize <- function(width = NULL,
                                height = NULL,
                                mapsize = NULL) {
  if (!is.null(width) && !is.null(height)) {
    mapsize <- c(width, height)
  }

  mapsize <- round(as.numeric(mapsize))

  stopifnot(
    is.numeric(mapsize) && (length(mapsize) == 2)
  )

  paste(mapsize, collapse = ",")
}

#' Convert numeric or character vector for orientation into dir query parameter
#' @noRd
get_dir_from_orientation <- function(orientation = NULL) {
  if (is.character(orientation)) {
    orientation <-
      match.arg(orientation, c("N", "E", "S", "W"))

    orientation <-
      switch(
        EXPR = orientation,
        "N" = 0,
        "E" = 90,
        "S" = 180,
        "W" = 270
      )
  }

  # FIXME: I'm unsure if orientation is required
  stopifnot(
    is.numeric(orientation)
  )

  if (orientation > 360) {
    orientation <-
      orientation - (360 * floor(orientation / 360))
  } else if (orientation < 0) {
    orientation <-
      orientation + (360 * ceiling(abs(orientation) / 360))
  }

  closest_cardinal_dir(orientation)
}

#' Get zoom query parameter
#' @noRd
get_zoom_for_imagery <- function(zoom = NULL, imagery = NULL) {
  if (is.null(zoom)) {
    zoom <- 18
  }

  if (imagery %in% imagery_options[5:6]) {
    reset_zoom <- NULL

    if (zoom < 18) {
      reset_zoom <- 18
    } else if (zoom > 22) {
      reset_zoom <- 22
    }

    if (!is.null(reset_zoom)) {
      zoom <- reset_zoom
      warning(
        paste0(
          imagery,
          " imagery only supports zoom levels between 18 and 22.",
          "\nSetting zoom to ", reset_zoom, "."
        )
      )
    }
  } else if (zoom < 0) {
    zoom <- 0
    warning("zoom must be between 0 and 22. Setting zoom to 0.")
  } else if (zoom > 22) {
    zoom <- 22
    warning("zoom must be between 0 and 22. Setting zoom to 22.")
  }

  round(as.numeric(zoom))
}

#' @rdname bing_static_map
#' @name get_request_url
#' @export
get_request_url <- function(...) {
  req_bingmapr(..., .perform = FALSE)$url
}

#' @return get_map_image returns an image from `magick::image_read`
#' @rdname bing_static_map
#' @name get_map_image
#' @export
#' @importFrom magick image_read
get_map_image <- function(location = NULL,
                          query = NULL,
                          imagery = "BirdsEye",
                          width = 600,
                          height = 400,
                          mapsize = NULL,
                          zoom = 18,
                          orientation = 0,
                          nudge = NULL,
                          key = Sys.getenv("BING_MAPS_API_KEY"),
                          check = TRUE,
                          strip = TRUE) {
  resp <-
    req_bingmapr(
      location = location,
      query = query,
      imagery = imagery,
      width = width,
      height = height,
      mapsize = mapsize,
      zoom = zoom,
      orientation = orientation,
      nudge = nudge,
      key = key,
      check = check,
      .perform = TRUE
    )

  magick::image_read(httr2::resp_body_raw(resp), strip = strip)
}


#' @return get_map_meta returns the JSON with the map metadata or a bbox for the
#'   map area
#' @rdname bing_static_map
#' @export
#' @importFrom httr2 resp_body_json
#' @importFrom sf st_bbox st_crs
get_map_meta <- function(location = NULL,
                         query = NULL,
                         imagery = "BirdsEye",
                         width = 600,
                         height = 400,
                         mapsize = NULL,
                         zoom = 18,
                         orientation = 0,
                         nudge = NULL,
                         key = Sys.getenv("BING_MAPS_API_KEY"),
                         bbox = FALSE) {
  resp <-
    req_bingmapr(
      location = location,
      query = query,
      imagery = imagery,
      width = width,
      height = height,
      mapsize = mapsize,
      zoom = zoom,
      orientation = orientation,
      nudge = nudge,
      key = key,
      mapMetadata = 1
    )

  meta <-
    httr2::resp_body_json(resp)

  if (bbox) {
    bbox <- sf::st_bbox(
      c(
        ymin = meta$resourceSets[[1]]$resources[[1]]$bbox[[1]],
        xmin = meta$resourceSets[[1]]$resources[[1]]$bbox[[2]],
        ymax = meta$resourceSets[[1]]$resources[[1]]$bbox[[3]],
        xmax = meta$resourceSets[[1]]$resources[[1]]$bbox[[4]]
      ),
      crs = sf::st_crs(4326)
    )

    return(bbox)
  }

  meta
}
